### ISSUE TITLE
`SimpleCacheFileSystem` ignores subfolder name if `same_names=True`, yielding invalid results

### ISSUE THREAD

**Issue by AndreiBarsan** [intake/filesystem_spec#735]
This causes invalid reads from cached data if you have `foo/file.txt` and `bar/file.txt` as the cache seems to be completely flat. Possibly related to #633, in that the file cache keying could be enhanced to include additional data, like the full path and, for S3 the version ID.

Here's a snippet that exemplifies the error:

```python
import fsspec
from fsspec.implementations.cached import SimpleCacheFileSystem


def main():
    base_fs = fsspec.filesystem("file")
    filesystem = SimpleCacheFileSystem(
        fs=base_fs, same_names=True, cache_storage="/tmp/fsspec-rfs-cache", cache_check=30
    )

    base_fs.mkdir("dog")
    base_fs.mkdir("cat")
    with filesystem.open("./dog/animal.txt", "w") as f:
        f.write("woof")
    with filesystem.open("./cat/animal.txt", "w") as f:
        f.write("meow")

    with filesystem.open("./dog/animal.txt", "r") as f:
        assert f.read() == "woof"
    with filesystem.open("./cat/animal.txt", "r") as f:
        assert f.read() == "meow"  # reads the wrong "animal.txt" in the cache - boom!


if __name__ == "__main__":
    main()
```

**Comment by martindurant** [comment]
Yes you are right, and this is the reason that `simple_names` is off by default. Without it, the final file name is generated by a hash on the whole URL. Also, it's not too obvious why you want to store files under their original names, if the intent is to access them via fsspec - you can let SimpleCache figure out which file is which.

We have no obvious place to store extra information about the file, as the whole point of Simple cache is to not have to update a metadata file. There have been some suggestions, perhaps adding a `.meta` file alongside every stored file, etc. However, it'd be good to come up with something comprehensive and configurable.

**Comment by AndreiBarsan** [comment]
Thank you for the quick reply! 

The reason is that the code needs the cached files to keep their original extension because of a library (sadly) beyond our control which refuses to read certain file formats unless given a path with the right extension. 

I see a few ways to possibly solve this:
 - (1) support hashed URI file names that still keep the original extension (so `s3://my_image.jpg` becomex `abc123.jpg` in the cache)
 - (2) document the fact that `same_names` ignores the path prefix, which is not immediately obvious from the API or from the docs

**Comment by martindurant** [comment]
I support both of those options. Another simple configuration addition may be: counting from the tail of a path, how many path elements to keep, with or without the hash.
Would you like to try implement some of these things? Obviously, some tests showing that they behave as intended would be appreciated.

**Comment by AndreiBarsan** [comment]
Here's something really coarse I just came up with:

```python

class HashMode:
    SAME_NAME = 1
    HASH = 2
    HASH_WITH_EXTENSION = 3


def hash_name(path: str, hash_mode: HashMode) -> str:
    if hash_mode == HashMode.SAME_NAME:
        return os.path.basename(path)
    elif hash_mode == HashMode.HASH_WITH_EXTENSION:
        hash = hashlib.sha256(path.encode()).hexdigest()
        _, ext = os.path.splitext(os.path.basename(path))
        return hash + "." + ext
    elif hash_mode == HashMode.HASH:
        return hashlib.sha256(path.encode()).hexdigest()
    else:
        raise ValueError(f"Unknown hash mode: {hash_mode}")
```

Still thinking about ways to make this work better with the existing API.

**Comment by martindurant** [comment]
Yes, something like that - but we would want to nicely deprecate the existing flags. Note that the naming of files, and how we store them, versus any metadata storage are all separate concerns, and it would be nice if, in the long run, we were able to pick and choose components for each.

**Comment by martindurant** [comment]
No rush from me, the current system has worked well enough for most for a while. Note that you can, of course, have different simple-cache instances with different cache locations, if you like. That is typical, for instance, in Intake data packages where the cached data sits alongside the catalogue that references it.
